### ISSUE TITLE
Default AssemblyLoadContext resolves assembly on the 2nd attempt

### ISSUE THREAD

**Issue by 2mik** [dotnet/runtime#74068]
### Description

Using .NET 6, I load DLL dynamically into the default load context by calling Assembly.LoadFrom(fileName)
The loaded assembly have some dependencies.
Using AssemblyResolve and AssemblyLoad events I got that 
1) AssemblyResolve shows that resolution of a referenced assembly fails.
2) Right after fail AssemblyLoad shows that the referenced assembly loaded successfully.

Please give a clue how to research the cause of that problem.

### Reproduction Steps

```
string fileName = "...";
string typeName = "...";
Assembly assembly = Assembly.LoadFrom(fileName);
Type type = assembly.GetType(typeName, true);
object obj = (ModuleLogic)Activator.CreateInstance(type);

AppDomain.CurrentDomain.AssemblyResolve += CurrentDomain_AssemblyResolve;
AppDomain.CurrentDomain.AssemblyLoad += CurrentDomain_AssemblyLoad;

// ...

private Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args)
{
            Console.WriteLine("!!! RequestingAssembly = " + args.RequestingAssembly.CodeBase);
            Console.WriteLine("    Item to resolve = " + args.Name);
            return null;
}

private void CurrentDomain_AssemblyLoad(object sender, AssemblyLoadEventArgs args)
{
            Console.WriteLine("!!! LoadedAssembly = " + args.LoadedAssembly.CodeBase);
}

```

### Expected behavior

The assembly is loaded successfully on the 1st attempt.
Of if assembly failed to load, it will not loaded later.

### Actual behavior

Assembly failed to load, then it is loaded successfully.

### Regression?

.NET Framework has another assembly loading mechanism, so it's not applicable. 

### Configuration

.NET 6, console application
Windows 11 x64

**Comment by 2mik** [comment]
The related issue may be https://github.com/dotnet/runtime/issues/1050

**Comment by elinor-fung** [comment]
> Assembly failed to load, then it is loaded successfully.

@2mik do you mean that you have two separate calls to `Assembly.LoadFrom` where the first one fails (throws), but the second one succeeds? The repro code in the description just tries to load once.

**Comment by elinor-fung** [comment]
Or is the concern here that your `AssemblyResolve` handler is called (and it returns null) and then your `AssemblyLoad` handler is called?

One thing to note with the `Assembly.LoadFrom` API is that it will [add a handler](https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/loading-managed#when-are-managed-assemblies-loaded) that will resolve assembly dependencies from the same directory as the assembly being loaded. If your `AssemblyResolve` handler was added first, it would be called before that handler for the `LoadFrom` (which would, if the dependency is next to the assembly, successfully resolve the dependency, resulting in the `AssemblyLoad` handler being called for it)

**Comment by 2mik** [comment]
> Or is the concern here that your `AssemblyResolve` handler is called (and it returns null) and then your `AssemblyLoad` handler is called?

Right. I called Assembly.LoadFrom once, but then get 2 events, AssemblyResolve event and AssemblyLoad event for the same DLL.
I added event handlers to AppDomain.CurrentDomain on application start. Note: I don't use `AssemblyLoadContext.Resolving` event because I need my code compatible with for .NET Standard.
Anyway, AssemblyLoadContext does not provide access to its internal resolvers. So I expect to handle an aggregated AssemblyResolve event if all built-in resolvers fails. Otherwise, the current behavior seems weird.

**Comment by 2mik** [comment]
@vitek-karas , thank you for the explanation. Now the situation is clear.
May be if I could see in debugger all the handlers which are called to resolve an assembly, it would be easy to understand by myself. Do you think it's possible to make handlers accessible from custom code in the future versions of .NET?

**Comment by vitek-karas** [comment]
The resolution algorithm is relatively complicated, so "handlers" is not the right way to think about it - you can read more here: https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/overview

If you need to diagnose subtle issues in the resolution process, we have tracing for that already: https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/collect-details

All of the interesting points (including all the extensions points) in the resolution process trace events, so collecting a trace like above should allow you to understand in detail what happened and when and what was the result of it.

Last option is - .NET is open source and for example VS can step into .NET source code (it will download the matching sources from github). So in theory you could debug all of it, but it's relatively tricky to do (the code goes between native and managed couple of times, which makes it difficult to step through).

**Comment by 2mik** [comment]
Thank you. The answer is an important addition to the mentioned documentation.
